File: 2040. Kth Smallest Product of Two Sorted Arrays.go
func kthSmallestProduct(nums1 []int, nums2 []int, k int64) int64 {
    l,r := -int64(1e10), int64(1e10)
    res := int64(0)
    for l <= r {
        mid := int64((l+r)/2)
        pairs := count(nums1, nums2, mid)
        if pairs < k {
            l = mid+1
        } else {
            res = mid
            r = mid - 1
        }
    }
     return res
}

func count(nums1 []int, nums2 []int, prod int64) int64 {
    pairs := 0
    for _,v := range nums1 {
        n := len(nums2)
        l := 0
        r := n - 1
        if v < 0 {
            m := n
            for l <= r {
                mid := ((l+r)/2)
                if int64(nums2[mid] * v) > prod {
                    l = mid + 1
                } else {
                    m = mid
                    r = mid - 1
                }
            } 
            pairs += (n-m)
        } else if v > 0 {
            m := -1
            for l <= r {
                mid := int((l+r)/2)
                if int64(nums2[mid] * v) > prod {
                    r = mid - 1
                } else {
                    m = mid
                    l = mid + 1
                }
            } 
            pairs += (m+1)
        } else {
            if prod >= 0{
                pairs += n
            }
        }
    }
    return int64(pairs)
}

File: 2040. Kth Smallest Product of Two Sorted Arrays.py
class Solution:
    def kthSmallestProduct(self, nums1: List[int], nums2: List[int], k: int) -> int:
        def countPairs(prod):
            pairs = 0 
            
            for num in nums1:
                
                l = 0
                r = len(nums2)-1
                if num < 0:
                    m = len(nums2)
                    while l<=r:
                        mid = (l+r)//2
                        if nums2[mid]*num > prod:
                            l = mid + 1
                        else:
                            m = mid
                            r = mid - 1 
                    pairs += len(nums2)- m
                elif num > 0:
                    m = -1
                    while l<=r:
                        mid = (l+r)//2
                        if nums2[mid]*num > prod:
                            r = mid - 1
                        else:
                            m = mid
                            l = mid + 1
                    pairs += (m + 1)
                else :
                    if prod >= 0:
                        pairs += len(nums2)
            return pairs
        
        l, r = -int(1e10), int(1e10)
        res = 0
        while l < r:
            mid = (l+r) // 2
            pairs = countPairs(mid)
            if pairs < k:
                l = mid + 1
            else:
                r = mid
        return r

        


